aws/rust-runtime/aws-runtime/src/sign.rs

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::auth;
+use aws_sigv4::http_request::{sign, SignableBody, SignableRequest, SigningParams};
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+
+/// Trait signing an HTTP request with a signature
+///
+/// # Examples
+/// ```rust,no_run
+/// use aws_runtime::sign::{SigningPackage, SignWith};
+/// use aws_sigv4::http_request::SigningParams;
+/// # use aws_smithy_runtime_api::box_error::BoxError;
+/// use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+///
+/// fn signing_params<'a>() -> SigningParams<'a> {
+///     // --snip--
+/// #    todo!()
+/// }
+///
+/// fn request() -> HttpRequest {
+///     // --snip--
+/// #   HttpRequest::empty()
+/// }
+///
+/// # fn example() -> Result<(), BoxError> {
+/// let signing_params = signing_params();
+/// let signing_package = SigningPackage::builder()
+///     .signing_params(&signing_params)
+///     .build()
+///     .expect("required fields have been set");
+///
+/// let mut request = request();
+/// request.sign_with(&signing_package)
+/// # }
+/// ```
+pub trait SignWith {
```

Comment:
What's your reasoning for putting this in aws-runtime instead of aws-sigv4?

Reply:
I was under the impression that moving forward, we'll be putting items in `aws-runtime-api` and `aws-runtime` crates, but not in constellation crates.

Reply:
I was under the impression that moving forward, we'll be putting items in `aws-runtime-api` and `aws-runtime` crates, but not in constellation crates, otherwise we can put it in `aws-sigv4` as well.

Reply:
I was under the impression that moving forward, we'll be putting items in `aws-runtime-api` and `aws-runtime` crates, but not in constellation crates, otherwise we can put it in `aws-sigv4` instead.

Reply:
`impl SignWith for HttpRequest` internally uses a module private `aws_runtime::auth::apply_signing_instructions`, so we need to move it to `aws-sigv4`.